lib/pbench/agent/tool_meister.py

```diff
@@ -1137,6 +1193,179 @@ def sysinfo(self, data):
         return failures
 
 
+def get_logger(PROG, daemon=False):
+    # If in Unit Test environment, just log to console
+    # If in non-unit test environment:
+    #   if daemonized, log to syslog and log back to Redis
+    #   if not daemonized, log to console and log back to Redis
+    # We assume
+    logger = logging.getLogger(PROG)
+    if os.environ.get("_PBENCH_TOOL_MEISTER_LOG_LEVEL") == "debug":
+        log_level = logging.DEBUG
+    else:
+        log_level = logging.INFO
+    logger.setLevel(log_level)
+
+    unit_tests = bool(os.environ.get("_PBENCH_UNIT_TESTS"))
+    if unit_tests or not daemon:
+        sh = logging.StreamHandler()
+    else:
+        sh = logging.SysLogHandler()
+    sh.setLevel(log_level)
+    shf = logging.Formatter(fmtstr_ut if unit_tests else fmtstr)
+    sh.setFormatter(shf)
+    logger.addHandler(sh)
+
+    return logger
+
+
+def driver(
+    PROG,
+    tar_path,
+    sysinfo_dump,
+    pbench_bin,
+    param_key,
+    params,
+    redis_server,
+    logger=None,
+):
+    """driver - responsible for actually creating the Tool Meister instance and
+    driving it operation.
+    """
+    if logger is None:
+        logger = get_logger(PROG)
+
+    # Add a logging handler to send logs back to the Redis server, with each
+    # log entry prepended with the given hostname parameter.
+    rh = RedisHandler(
+        channel="tm-logging", hostname=params["hostname"], redis_client=redis_server
+    )
+    if os.environ.get("_PBENCH_TOOL_MEISTER_LOG_LEVEL") == "debug":
```

Comment:
Okay, this is ugly.  We call `get_logger()` from two places to handle daemonized vs non-daemonized.  I admit, I did not spend my time thinking about how to make this cleaner.